winpr/libwinpr/utils/wlog/wlog.c

```diff
@@ -66,12 +66,99 @@ LPCSTR WLOG_LEVELS[7] =
 	"OFF"
 };
 
+static INIT_ONCE _WLogInitialized = INIT_ONCE_STATIC_INIT;
 static DWORD g_FilterCount = 0;
 static wLogFilter* g_Filters = NULL;
+static wLog* g_RootLog = NULL;
 
+static wLog* WLog_New(LPCSTR name, wLog* rootLogger);
+static void WLog_Free(wLog* log);
 static LONG WLog_GetFilterLogLevel(wLog* log);
 static int WLog_ParseLogLevel(LPCSTR level);
 static BOOL WLog_ParseFilter(wLogFilter* filter, LPCSTR name);
+static BOOL WLog_ParseFilters(void);
+
+static void WLog_Uninit_(void)
+{
+	DWORD index;
+	wLog* child = NULL;
+	wLog* root = g_RootLog;
+
+	if (!root)
+		return;
+
+	for (index = 0; index < root->ChildrenCount; index++)
+	{
+		child = root->Children[index];
+		WLog_Free(child);
+	}
+
+	WLog_Free(root);
+	g_RootLog = NULL;
+}
+
+static BOOL CALLBACK WLog_InitializeRoot(PINIT_ONCE InitOnce, PVOID Parameter, PVOID* Context)
+{
+	char* env;
+	DWORD nSize;
+	DWORD logAppenderType;
+	LPCSTR appender = "WLOG_APPENDER";
+
+	if (!(g_RootLog = WLog_New("", NULL)))
+		return FALSE;
+
+	g_RootLog->IsRoot = TRUE;
+	WLog_ParseFilters();
+	logAppenderType = WLOG_APPENDER_CONSOLE;
+	nSize = GetEnvironmentVariableA(appender, NULL, 0);
+
+	if (nSize)
+	{
+		env = (LPSTR) malloc(nSize);
+
+		if (!env)
+			goto fail;
+
+		if (GetEnvironmentVariableA(appender, env, nSize) != nSize - 1)
+		{
+			fprintf(stderr, "%s environment variable modified in my back", appender);
+			free(env);
+			goto fail;
+		}
+
+		if (_stricmp(env, "CONSOLE") == 0)
+			logAppenderType = WLOG_APPENDER_CONSOLE;
+		else if (_stricmp(env, "FILE") == 0)
+			logAppenderType = WLOG_APPENDER_FILE;
+		else if (_stricmp(env, "BINARY") == 0)
+			logAppenderType = WLOG_APPENDER_BINARY;
+
+#ifdef HAVE_SYSLOG_H
+		else if (_stricmp(env, "SYSLOG") == 0)
+			logAppenderType = WLOG_APPENDER_SYSLOG;
+
+#endif /* HAVE_SYSLOG_H */
+#ifdef HAVE_JOURNALD_H
+		else if (_stricmp(env, "JOURNALD") == 0)
+			logAppenderType = WLOG_APPENDER_JOURNALD;
+
+#endif
+		else if (_stricmp(env, "UDP") == 0)
+			logAppenderType = WLOG_APPENDER_UDP;
+
+		free(env);
+	}
+
+	if (!WLog_SetLogAppenderType(g_RootLog, logAppenderType))
+		goto fail;
+
+	atexit(WLog_Uninit_);
```

Comment:
I think this could get us into trouble. MSDN says:

"The code in the atexit function should not contain any dependency on any DLL which could have already been unloaded when the atexit function is called."

If someone uses a custom logger with cleanup code like i.e. WLog_UdpAdapter_Free (which calls socket APIs) this could lead to crashes on app exit.

Not sure what you guys think about this - maybe I'm just too pessimistic here ... ^^

Reply:
@mfleisz That would also crash if `WLog_*` functions are called...
Don't think that the dependent libs are already unloaded there (unless someone really wants to get hurt)

Reply:
@akallabeth WLog_* functions won't be a problem as they are part of winpr.dll and are still loaded when the atexit() handler is called (on dll unload of winpr). There is just no guarantee the adapter free handlers won't call any functions from other dlls that might have been already unloaded before winpr.

Reply:
@mfleisz My understanding of this is that it is only relevant when using `WinPR` with `dlopen`/`dlclose` (or `LoadLibrary`/`CloseHandle`)